drivers/dma/dma_xilinx_axi_dma.c

```diff
@@ -0,0 +1,1020 @@
+		k_oops();
+		return -ENOTSUP;
+	}
+
+	if (dma_cfg->head_block->dest_addr_adj == DMA_ADDR_ADJ_DECREMENT) {
+		LOG_ERR("Xilinx AXI DMA only supports incrementing addresses!");
+		k_oops();
+		return -ENOTSUP;
+	}
+
+	if (dma_cfg->head_block->source_addr_adj != DMA_ADDR_ADJ_INCREMENT &&
+	    dma_cfg->head_block->source_addr_adj != DMA_ADDR_ADJ_NO_CHANGE) {
+		LOG_ERR("invalid source_addr_adj %" PRIu16, dma_cfg->head_block->source_addr_adj);
+		k_oops();
+		return -ENOTSUP;
+	}
+	if (dma_cfg->head_block->dest_addr_adj != DMA_ADDR_ADJ_INCREMENT &&
+	    dma_cfg->head_block->dest_addr_adj != DMA_ADDR_ADJ_NO_CHANGE) {
+		LOG_ERR("invalid dest_addr_adj %" PRIu16, dma_cfg->head_block->dest_addr_adj);
+		k_oops();
+		return -ENOTSUP;
+	}
+
+	if (channel == XILINX_AXI_DMA_TX_CHANNEL_NUM &&
+	    dma_cfg->channel_direction != MEMORY_TO_PERIPHERAL) {
+		LOG_ERR("TX channel must be used with MEMORY_TO_PERIPHERAL!");
+		k_oops();
+		return -ENOTSUP;
+	}
+
+	if (channel == XILINX_AXI_DMA_RX_CHANNEL_NUM &&
+	    dma_cfg->channel_direction != PERIPHERAL_TO_MEMORY) {
+		LOG_ERR("RX channel must be used with PERIPHERAL_TO_MEMORY!");
+		k_oops();
+		return -ENOTSUP;
+	}
+
+	data->channels[channel].last_transfer_direction = dma_cfg->channel_direction;
+
+	dma_xilinx_axi_dma_disable_cache();
+
+	if (channel == XILINX_AXI_DMA_TX_CHANNEL_NUM) {
+		data->channels[channel].descriptors = descriptors_tx;
+		data->channels[channel].num_descriptors = ARRAY_SIZE(descriptors_tx);
+
+		data->channels[channel].channel_regs = &regs->mm2s_registers;
+	} else {
+		data->channels[channel].descriptors = descriptors_rx;
+		data->channels[channel].num_descriptors = ARRAY_SIZE(descriptors_rx);
+
+		data->channels[channel].channel_regs = &regs->s2mm_registers;
+	}
+
+	LOG_DBG("Resetting DMA channel!");
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Waddress-of-packed-member"
+	dma_xilinx_axi_dma_write_reg(&data->channels[channel].channel_regs->dmacr,
+				     XILINX_AXI_DMA_REGS_DMACR_RESET);
+#pragma GCC diagnostic pop
+
+	LOG_DBG("Configuring %zu DMA descriptors for %s", data->channels[channel].num_descriptors,
+		channel == XILINX_AXI_DMA_TX_CHANNEL_NUM ? "TX" : "RX");
+
+	/* only configures fields whos default is not 0, as descriptors are in zero-initialized */
+	/* segment */
+	data->channels[channel].current_transfer_start_index =
+		data->channels[channel].current_transfer_end_index = 0;
+	for (int i = 0; i < data->channels[channel].num_descriptors; i++) {
+		uintptr_t nextdesc;
+		uint32_t low_bytes;
+#ifdef CONFIG_DMA_64BIT
+		uint32_t high_bytes;
+#endif
+		if (i + 1 < data->channels[channel].num_descriptors) {
+			nextdesc = (uintptr_t)&data->channels[channel].descriptors[i + 1];
+		} else {
+			nextdesc = (uintptr_t)&data->channels[channel].descriptors[0];
+		}
+		/* SG descriptors have 64-byte alignment requirements */
+		/* we check this here, were each descriptor */
+		if (nextdesc & XILINX_AXI_DMA_SG_DESCRIPTOR_ADDRESS_MASK) {
```

Comment:
__ASSERT() instead, this is a programming error

Reply:
Done